swarm/swap_test.go

```diff
@@ -0,0 +1,423 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swarm
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"math/rand"
+	"os"
+	"strconv"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethereum/go-ethereum/p2p/simulations/adapters"
+	"github.com/ethereum/go-ethereum/swarm/api"
+	"github.com/ethereum/go-ethereum/swarm/log"
+	"github.com/ethereum/go-ethereum/swarm/network/simulation"
+	"github.com/ethereum/go-ethereum/swarm/storage"
+)
+
+var (
+	printStats     = flag.Bool("printstats", false, "print swap stats")
+	bucketKeySwarm = simulation.BucketKey("swarm")
+)
+
+//In TestSwapNetworkSymmetricFileUpload we set up a network with arbitrary number of nodes
+//(16), and each of the nodes uploads a file of same size
+//Afterwards we check that every node's balance WITH ANOTHER PEER
+//has the same value but opposite sign
+func TestSwapNetworkSymmetricFileUpload(t *testing.T) {
```

Comment:
I dont see why we would need symmetric and asymmetric versions of this test. What you check is the same no matter how many and what files a node retrieves. Why not just use the exact same simulation test as we have in accounting. 
In fact adding this test has not much additional value, unless you check the exact amounts as well. 

The exact amounts should be tested however with much simpler unit tests for a particular exchange

Reply:
> I dont see why we would need symmetric and asymmetric versions of this test. What you check is the same no matter how many and what files a node retrieves. 

Well yes theoretically. But I was unsure this was indeed the case and wanted proof. 

> The exact amounts should be tested however with much simpler unit tests for a particular exchange
I only partly agree. I will never resist more and simpler unit tests - the more, the better. Nevertheless we don't have any end-to-end tests of the complete stack concerning accounting, involving actual file transfers and accounting for them. So I maintain this simulation to be very valuable.

I do think though that the test may be moved into a set of end-to-end tests - which we don't have any infrastructure and setup yet. So we wouldn't have such IMHO important functionality tests which are not strictly unit test adding up to the automated test runs (prolonging test duration).

But until we don't have such infrastructure (where I have long advocated we should move the snapshot sync tests to), I would very much like these kind of tests be kept - until proven wrong of course.

Reply:
They are not affected. These tests are at the streamer protocol level, so they just assume peers have been correctly established.

I insist we can not ditch e2e just because they are not unit tests. When we have an established way to run and maintain e2e tests, then I will certainly agree to move tests like these there. But for the time being, this test confirms to me that uploading real files results in correct balances.

Reply:
Unit tests would in any case only account one way, to really check that the balances are working correctly for the high level file API you need message exchanges and multiple nodes and thus e2e tests.

Reply:
I removed all e2e tests